File: swim-java/swim-core-java/swim.structure/src/main/java/swim/structure/DataOutput.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.structure;

import swim.codec.Output;
import swim.codec.OutputSettings;

final class DataOutput extends Output<Data> {
  final Data data;
  OutputSettings settings;

  DataOutput(Data data, OutputSettings settings) {
    this.data = data;
    this.settings = settings;
  }

  @Override
  public boolean isCont() {
    return true;
  }

  @Override
  public boolean isFull() {
    return false;
  }

  @Override
  public boolean isDone() {
    return false;
  }

  @Override
  public boolean isError() {
    return false;
  }

  @Override
  public boolean isPart() {
    return false;
  }

  @Override
  public Output<Data> isPart(boolean isPart) {
    return this;
  }

  @Override
  public Output<Data> write(int b) {
    this.data.addByte((byte) b);
    return this;
  }

  @Override
  public Output<Data> write(String string) {
    throw new UnsupportedOperationException();
  }

  @Override
  public Output<Data> writeln(String string) {
    throw new UnsupportedOperationException();
  }

  @Override
  public Output<Data> writeln() {
    throw new UnsupportedOperationException();
  }

  @Override
  public OutputSettings settings() {
    return this.settings;
  }

  @Override
  public Output<Data> settings(OutputSettings settings) {
    this.settings = settings;
    return this;
  }

  @Override
  public Data bind() {
    return this.data;
  }

  @Override
  public Output<Data> clone() {
    return new DataOutput(this.data.branch(), this.settings);
  }
}

File: swim-js/swim-runtime/swim-core/swim-structure/src/main/DataOutput.ts
// Copyright 2015-2023 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type {Mutable} from "@swim/util";
import {AnyOutputSettings, OutputSettings, Output} from "@swim/codec";
import type {Data} from "./Data";

/** @internal */
export class DataOutput extends Output<Data> {
  constructor(data: Data, settings: OutputSettings) {
    super();
    this.data = data;
    this.settings = settings;
  }

  /** @internal */
  readonly data: Data;

  override isCont(): boolean {
    return true;
  }

  override isFull(): boolean {
    return false;
  }

  override isDone(): boolean {
    return false;
  }

  override isError(): boolean {
    return false;
  }

  override isPart(): boolean {
    return false;
  }

  override asPart(part: boolean): Output<Data> {
    return this;
  }

  override write(b: number | string): Output<Data> {
    if (typeof b === "number") {
      this.data.addByte(b);
      return this;
    } else {
      throw new TypeError("" + b);
    }
  }

  override writeln(string?: string): Output<Data> {
    throw new TypeError("" + string);
  }

  override readonly settings: OutputSettings;

  override withSettings(settings: AnyOutputSettings): Output<Data> {
    settings = OutputSettings.fromAny(settings);
    (this as Mutable<this>).settings = settings;
    return this;
  }

  override bind(): Data {
    return this.data;
  }

  override clone(): Output<Data> {
    return new DataOutput(this.data.branch(), this.settings);
  }
}
